### ISSUE TITLE
[FR] Allow for warm up before running each benchmark

### ISSUE THREAD

**Issue by Adrodoc** [google/benchmark#1130]
**Is your feature request related to a problem? Please describe.**
Some code is slow at first but becomes faster after some time. In my example I am benchmarking an MPI application and the first few benchmark iterations run a lot slower (I suspect because the memory of the MPI-Window is not fully initialized and put into cache).
For example here are the timings of the first 10 iterations of some benchmark I am working with:
```
seconds=0.0433436
seconds=0.0381923
seconds=0.0215415
seconds=0.0117373
seconds=0.0098229
seconds=0.0090367
seconds=0.0100212
seconds=0.0068646
seconds=0.0071745
seconds=0.0073567
```
As you can see it takes a few iterations to warm up and then it stays at about 0.007 seconds which is almost 6 times faster than the initial iteration.

**Describe the solution you'd like**
I would like to specify a `MinWarmUpTime` that acts like `MinTime` to dynamically determine a number of iterations to run a warm up. All timings of the warm up should be discarded (to not impact the reported average) and they should not be used to determine the number of iterations for the actual benchmark.
After the warm up the iteration count of the benchmark itself would be determined using the next few iterations as normal using `MinTime`.

**Describe alternatives you've considered**
You could also specify a fixed iteration count for warm up, but then users would have to first find a good count manually (this is what I am doing now).
You could also specify a fixed `WarmUpTime` and always run the warm up for that time. This might be easier than dynamically determine a number of iterations and should work fine too.

**Comment by Matthdonau** [comment]
Can be closed i guess?

**Comment by dominichamon** [comment]
Fixed by #1399
